### PR TITLE
Server adjustments

### PR DESCRIPTION
**Pull request by robhenley**
Here are some minor adjustments to the server including:
- An API specific subrouter
- Middleware for logging, CORS, and setting a default content type
- An environment variable to set the servers address w/ feedback the server is running
- Slack oAuth2 authentication

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -95,14 +113,116 @@ func enableCors(w *http.ResponseWriter) {
 
 func handleRequest() {
 
+	serverAddr, ok := os.LookupEnv("JOBS_SERVER_ADDR")
+	if !ok {
+		serverAddr = ":8081"
```
Comment (ShaunWMusic) [inline main.go:118]: this is just opinion, but i think it reads better

```suggestion
	serverAddr, err := os.LookupEnv("JOBS_SERVER_ADDR")
	if err != nil {
		serverAddr = ":8081"
```

Comment (ShaunWMusic) [reply main.go:118]: wait... did i understand that correctly?

File: main.go [hunk 2]
```diff
@@ -1,18 +1,36 @@
...
 
+// Move this somewhere more appropriate
+var (
+	authConfig = &oauth2.Config{
+		RedirectURL:  os.Getenv("ORIGIN") + "/auth/callback",
+		ClientID:     os.Getenv("SLACK_CLIENT_ID"),
+		ClientSecret: os.Getenv("SLACK_CLIENT_SECRET"),
+		Scopes:       []string{"identity.basic"},
+		Endpoint:     endpoints.Slack,
+	}
```
Comment (ShaunWMusic) [inline main.go:30]: nice! is this working with slack?

File: migrations/20210305124354_create_initial_tables.down.sql [hunk 3]
```diff
@@ -0,0 +1,4 @@
+DROP TABLE companies;
+DROP TABLE jobs;
+DROP TABLE social_accounts;
+DROP TABLE users;
```
Comment (ShaunWMusic) [inline migrations/20210305124354_create_initial_tables.down.sql]: 👍

Comment (ShaunWMusic) [reply migrations/20210305124354_create_initial_tables.down.sql]: have you successfully ran `up` and `down` migrations?

File: main.go [hunk 4]
```diff
@@ -95,14 +113,116 @@ func enableCors(w *http.ResponseWriter) {
...
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Add("Content-Type", jsonapi.MediaType)
+		next.ServeHTTP(w, r)
+	})
+}
+
+func getBasicInfo(accessToken string) (d map[string]interface{}, err error) {
+
+	// NOTE: DO NOT USE https://slack.com/api/auth.test as it always
+	// responds with "invalid_auth".  I'm guessing it's a legacy endpoint
```
Comment (ShaunWMusic) [inline main.go:206]: ah ok, so is it working with slack?

File: main.go (hunk 1 above)
Comment (robhenley) [reply main.go:118]: Just to recap our conversation in Slack my understanding is `ok` is idiomatic Go for `bool` types and err is for `error` types with [`os.LookupEnv`](https://golang.org/pkg/os/#LookupEnv) returning a `string` and `bool` types respectively `ok` still makes sense to me but I'm happy to change.

File: main.go (hunk 2 above)
Comment (robhenley) [reply main.go:30]: Yes, but just the log in part.  It's not storing in a session or anything...

File: migrations/20210305124354_create_initial_tables.up.sql [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    expires_in bigint,
+    created_at date,
+    updated_at date,
+    PRIMARY KEY (id),
+    CONSTRAINT social_accounts_user_id_foreign FOREIGN KEY (user_id)
+        REFERENCES users (id) MATCH SIMPLE
+        ON UPDATE CASCADE
+        ON DELETE CASCADE
+        NOT VALID
+);
```
Comment (ShaunWMusic) [inline migrations/20210305124354_create_initial_tables.up.sql]: i'm not sure we actually need this... do we?

File: migrations/README.md [hunk 6]
```diff
@@ -0,0 +1,9 @@
+# DB Migrations
+
+SQL migrations are done with [Golang Migrate](https://github.com/golang-migrate/migrate) installed with `brew install golang-migrate`.
+
+You can create a migration with:
+`migrate create -dir migrations/ -ext sql <migration_name>`
+
+You can run the migrations with for example:
+`migrate -source file://migrations -database 'postgres://<user>:<password>@127.0.0.1:5432/<database>?sslmode=disable' up`
```
Comment (ShaunWMusic) [inline migrations/README.md:9]: nice! more docs the better!

Comment (ShaunWMusic) [reply migrations/README.md:9]: i'd like to install swagger docs as well at some point

**Review by ShaunWMusic — APPROVED** [review]
approved, i left some comments and questions. but overall it's looking good.

File: migrations/20210305124354_create_initial_tables.down.sql (hunk 3 above)
Comment (robhenley) [reply migrations/20210305124354_create_initial_tables.down.sql]: Yes, but the migrations need some work.  The user id returned from Slack is a string and it sounded like you were thinking we didn't need a `users` table.  I didn't intend to include the migrations in this pull request but was working in a hurry and branched from the previous PR.

File: main.go (hunk 4 above)
Comment (robhenley) [reply main.go:206]: Yes, I put that as a warning for myself (or another dev).  That endpoint returns an "invalid_auth" even with a valid token.  I sent a note to Slack but haven't heard anything.  There were some comments online about them switching token types so I'm assume it's for the legacy token but couldn't find any confirmation.  I confirm the token is working in `getBasicInfo` function.

File: migrations/20210305124354_create_initial_tables.up.sql (hunk 5 above)
Comment (robhenley) [reply migrations/20210305124354_create_initial_tables.up.sql]: My thought was supporting multiple auth providers but if we're only ever going to use Slack and we use the token to post to the Slack group then I agree.

Comment (robhenley) [reply migrations/20210305124354_create_initial_tables.up.sql]: I looked at the Slack API some more and it doesn't look like we need a user token at all even if we want to support posting the job automatically as that can be done either as the user (requiring the user token) or the bot.  I looked and it looks like NashDev is using the "bot" token... thoughts?

Comment (ShaunWMusic) [reply migrations/20210305124354_create_initial_tables.up.sql]: nice! i'm good with using the "bot" token

Comment (ShaunWMusic) [reply migrations/20210305124354_create_initial_tables.up.sql]: i don't see us supporting multiple auth providers

Comment (robhenley) [reply migrations/20210305124354_create_initial_tables.up.sql]: Yea I started to adjust this but just haven't had time to circle back.  My original plan is to remove the social_accounts table and add some of the Slack details (user_id, access_token, team) to the users table.  For the current plan I'm not sure any of those are strictly necessary, though.  Also to switch to using Postgres's `SERIAL` or `BIGSERIAL`.

**Comment by robhenley** [comment]
@ShaunWMusic This should be ready to merge.

**Pull request merged**